pkg/edgeview/dispatcher/wss-server.go

```diff
@@ -0,0 +1,265 @@
+		for addr, e := range tmpMap {
+			if e.hostname == hostname {
+				fmt.Printf("%v received connection with same hostname %s, close old w/Addr %s\n", time.Now(), hostname, addr)
+				e.wsConn.Close()
+				addOK = true
+			}
+		}
+		if !addOK {
+			err := conn.WriteMessage(websocket.TextMessage, []byte(moretwoMsg))
+			if needDebug {
+				fmt.Printf("websocket write: %v\n", err)
+			}
+			connMutex.Unlock()
+			return
+		}
+	}
+
+	ep := endPoint{
+		wsConn:    conn,
+		hostname:  hostname,
+	}
+	if _, ok := reqAddrTokenEP[token][remoteAddr]; !ok {
+		reqAddrTokenEP[token][remoteAddr] = ep
+	}
+	sizeMap := len(tmpMap)
+	connMutex.Unlock()
+	if sizeMap < 2 {
+		err := conn.WriteMessage(websocket.TextMessage, []byte(noDeviceMsg))
+		if needDebug {
+			fmt.Printf("websocket write: %v\n", err)
+		}
+	}
+	fmt.Printf("%v client %s from %s connected, ID: %d\n",
+		time.Now().Format("2006-01-02 15:04:05"), hostname, remoteAddr, myConnID)
+
+	cnt := 0
+	nopeerPkts := 0
+    for {
+        messageType, message, err := conn.ReadMessage()
+		now := time.Now()
+		nowStr := now.Format("2006-01-02 15:04:05")
+        if err != nil {
+            fmt.Printf("%s on reading host %s from %s, ID %d: %v\n", nowStr, hostname, remoteAddr, myConnID, err)
+			cleanConnMap(token, remoteAddr)
+            break
+        }
+
+		connMutex.Lock()
+		tmpMap = reqAddrTokenEP[token]
+		if tmpMap == nil {
+			connMutex.Unlock()
+			continue
+		}
+
+		myEP := endPoint{}
+		var peerAddr string
+
+		for addr, e := range tmpMap {
+			if remoteAddr == addr {
+				continue
+			}
+			dest := strings.Split(addr, ":")
+			if len(dest) == 2 {
+				addr = dest[1]
+			}
+			if needDebug {
+				fmt.Printf("%s (%d/%d): [%v], t-%d len %d, to %s\n",
+					nowStr, myConnID, cnt, hostname, messageType, len(message), addr)
+			}
+			peerAddr = addr
+			myEP = e
+			nopeerPkts = 0
+			break
+		}
+		connMutex.Unlock()
+
+		if myEP.wsConn == nil {
+			nopeerPkts++
+			fmt.Printf("%s can not find peer %d\n", nowStr, nopeerPkts)
+			if nopeerPkts < 50 { // need sometime for ep to reconnect
+				continue
+			}
+			err = conn.WriteMessage(websocket.TextMessage, []byte(noDeviceMsg))
+			if err != nil {
+				fmt.Printf("Error during message writing: %v\n", err)
+				cleanConnMap(token, remoteAddr)
+				break
+			}
+			continue
+		}
+		err = myEP.wsConn.WriteMessage(messageType, message)
+		if err != nil {
+			fmt.Printf("Error during message from %s writing to %s, ID %d: %v\n", hostname, peerAddr, myConnID, err)
+			cleanConnMap(token, remoteAddr)
+			break
+		}
+		cnt++
+    }
+}
+
+func cleanConnMap(token, remoteAddr string) {
+	connMutex.Lock()
+	tmpMap := reqAddrTokenEP[token]
+	if tmpMap != nil {
+		delete(tmpMap, remoteAddr)
+		if len(tmpMap) == 0 {
+			delete(reqAddrTokenEP, token)
+		}
+	}
+	connMutex.Unlock()
+}
+
+// Get preferred outbound ip of this machine
+func getOutboundIP() string {
+    conn, err := net.Dial("udp", "8.8.8.8:80")
+    if err != nil {
+        log.Fatal(err)
```

Comment:
In case of connectivity issues seems reasonable to try again instead of fatal here.

Reply:
sure.